staging/src/k8s.io/apiserver/pkg/util/flowcontrol/metrics/sample_and_watermark_test.go

```diff
@@ -0,0 +1,110 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"fmt"
+	"math/rand"
+	"testing"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/clock"
+	compbasemetrics "k8s.io/component-base/metrics"
+	"k8s.io/component-base/metrics/legacyregistry"
+	"k8s.io/klog/v2"
+)
+
+const (
+	samplesHistName = "sawtestsamples"
+	ddtRange        = 3000
+	ddtOffset       = 500
+	numIterations   = 100
+)
+
+/* TestSampler does a rough behavioral test of SampleAndWatermarkHistograms.
+   The test creates one and exercises it, checking that the count in the
+   sampling histogram is correct at each step.  A fake clock is used, and
+   the exercise consists of changing that fake clock by an amount of time
+   chosen uniformly at random from a range that goes from a little negative
+   to somewhat more than two sampling periods.  The negative changes are
+   included because negative changes have been observed in real monotonic
+   clock readings.
+*/
+func TestSampler(t *testing.T) {
+	t0 := time.Now()
+	clk := clock.NewFakePassiveClock(t0)
+	buckets := []float64{0, 1}
+	const samplingPeriod = time.Millisecond
+	gen := NewSampleAndWaterMarkHistogramsGenerator(clk, samplingPeriod,
+		&compbasemetrics.HistogramOpts{Name: samplesHistName, Buckets: buckets},
+		&compbasemetrics.HistogramOpts{Name: "marks", Buckets: buckets},
+		[]string{})
+	saw := gen.Generate(0, 1, []string{})
+	regs := gen.metrics()
+	for _, reg := range regs {
+		legacyregistry.MustRegister(reg)
+	}
+	dt := 2 * samplingPeriod
+	t1 := t0.Add(dt)
+	klog.Infof("Expect about %v warnings about time going backwards; this is fake time deliberately misbehaving.", (numIterations*ddtOffset)/ddtRange)
+	t.Logf("t0=%s", t0)
+	for i := 0; i < numIterations; i++ {
+		ddt := time.Microsecond * time.Duration(rand.Intn(ddtRange)-ddtOffset)
+		t1 = t1.Add(ddt)
+		diff := t1.Sub(t0)
+		if diff > dt {
+			dt = diff
+		}
+		clk.SetTime(t1)
+		saw.Set(1)
```

Comment:
Do you need to test other timeObserver operation like `Add` or `SetX1`?

Reply:
Additional tests could be created for those.   This one is explicitly and exclusively focused on the question of when samples are taken.